### PR TITLE
Feature/fir 21173 add identity support

### PR DESCRIPTION
**Pull request by alexradzin**
updated minimal implementation of the new identity

### CODE REVIEW

File: src/main/java/com/firebolt/jdbc/service/FireboltEngineService.java
Comment (ptiurin) [inline src/main/java/com/firebolt/jdbc/service/FireboltEngineService.java:33]: I believe the concept of a default engine will go away soon. The logic agreed at the moment is if no engine name is specified by the user - connect to a system engine.

File: src/test/java/com/firebolt/jdbc/client/query/StatementClientImplTest.java
Comment (ptiurin) [inline src/test/java/com/firebolt/jdbc/client/query/StatementClientImplTest.java:45]: Which engine url is unavailable here?

File: src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java
Comment (ptiurin) [inline src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java:402]: engine=system is no longer a valid parameter. The new workflow is if no engine name is specified the driver connects to a system engine.

File: src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java
Comment (ptiurin) [inline src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java:120]: I don't think this is correct. When db is not provided (but a user engine is) we fetch it via the system engine. If no engine and no db is provided we connect to a system engine with no db context.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java
Comment (ptiurin) [inline src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java:58]: We should be able to determine this value from the `HOST`, right? e.g. api.app.firebolt.io -> env=app

File: src/main/java/com/firebolt/jdbc/connection/UrlUtil.java
Comment (ptiurin) [inline src/main/java/com/firebolt/jdbc/connection/UrlUtil.java:25]: Is there no standard url parser we can use? Why do we have a custom parser here?

File: src/integrationTest/java/integration/tests/TimestampTest.java
Comment (ptiurin) [inline src/integrationTest/java/integration/tests/TimestampTest.java:144]: I don't think this test belongs here. This class deals with testing the Timestamp types.

File: src/main/java/com/firebolt/FireboltDriver.java
Comment (alexradzin) [reply src/main/java/com/firebolt/FireboltDriver.java:20]: done

File: src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java:209]: done

File: src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java:81]: Done. Although IMHO writing ugly code is worse than explaining why this ugly code is written.

File: src/main/java/com/firebolt/jdbc/client/account/FireboltAccountRetriever.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/client/account/FireboltAccountRetriever.java:14]: Added `env`  - hopefully it is good enough.

File: src/main/java/com/firebolt/jdbc/connection/UrlUtil.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/UrlUtil.java:25]: This code was written on 22.03.2023; I have not changed code I do not have to change. Also unfortunately JDK does not provide function that parses URL parameters and return query parameters as properites, map or something like this.

File: src/integrationTest/java/integration/tests/TimestampTest.java
Comment (alexradzin) [reply src/integrationTest/java/integration/tests/TimestampTest.java:144]: Although this test exists since March I am moving to to special test case.

File: src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java
Comment (alexradzin) [reply src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java:120]: Let's review test later as a part of general refactoring and the code face-lifting.

File: src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java
Comment (alexradzin) [reply src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java:402]: Let's disucuss this an fix the test together. I am not sure I understand what should be done here. Anyway this is an old test; I just improved the code style.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java:58]: Hopefully fixed. Please take a look on `FireboltPropertiesTest.hostAndEnvironment()` and `FireboltProperties.getEnvironment()`

File: src/main/java/com/firebolt/jdbc/client/account/FireboltAccountRetriever.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/client/account/FireboltAccountRetriever.java:14]: fixed

File: src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java
Comment (ptiurin) [reply src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java:208]: Still showing in the diff

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java
Comment (ptiurin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltSessionProperty.java:58]: Yeah, looks good. Having both environment and host is somewhat redundant, since we duplicate the info, but as long as it makes the interaction with the driver easier for us, why not.

File: src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java:208]: It is  shown as removed. Go to tab "Files Changed"

Comment (ptiurin) [reply src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java:208]: image

File: src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java
Comment (ptiurin) [reply src/test/java/com/firebolt/jdbc/connection/settings/FireboltPropertiesTest.java:122]: Clarified over the call, this is testing the Properties class and not the logic itself.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:117]: Generally null is bad. It can produce NullPointerException. Surely, it is possible but it is a little bit dangerous, so I'd suggest to postpone this. I'd change API to return Optional<String> instead but it is too much for now IMHO.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:133]: It is possible, but right now I discover environemnt from host only if environment was not supplied. This approach allows to override environment if it is needed.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:21]: Unfortunately it is still in use in a couple of palces and I do not want to use null instead.

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:59]: done

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (stepansergeevitch) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:24]: We would need to remove any mention of the term `system` customer-exposed code.
Also, this constant is related to functionality that no longer works so we need to make sure we don't rely on it
Before new identity, we used to pass `system` as the engine_name in order to create a connection to a system engine.
For the new identity, this will simply not work. In case we still use this constant in other places, it can mean that we still rely on/test this functionality somewhere, which is invalid and should be fixed. I think we would need to remove it anyways

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (stepansergeevitch) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:119]: As mentioned above, we would need to remove any mention of the term `system engine`, so I would suggest to simply return an empty string here.

File: .github/workflows/integration-test.yml
Comment (alexradzin) [reply .github/workflows/integration-test.yml:15]: Account is not the thing that should be stored in secret that it typically used for passwords and things like them that should not be visible. We can use hard coded value but I really do not understand why. Right now the account can be configured and its default value is "integration", so in most cases we do not have to change anything.

File: src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java
Comment (alexradzin) [reply src/test/java/com/firebolt/jdbc/connection/FireboltConnectionTest.java]: Done. No more "system".

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java]: Now this function is extremely simple. It returns engine as is. For system engine it returns `null`.

File: src/test/java/com/firebolt/jdbc/client/query/StatementClientImplTest.java
Comment (alexradzin) [reply src/test/java/com/firebolt/jdbc/client/query/StatementClientImplTest.java]: Everything is available and this test is enabled now.

File: src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/client/query/StatementClientImpl.java]: done

File: src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/settings/FireboltProperties.java:136]: `getEnvironment()` now implements this logic

File: src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java
Comment (stepansergeevitch) [inline src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java:145]: Just a question, should't it also be `throws FireboltException`, since we do `throw new FireboltException(...)`?

**Review by stepansergeevitch — APPROVED** [review]
Generally LGTM, left one question. I would also ask to wait for an approval from Petro

File: src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java:145]: No. Generally the interfaces should use general terms. Prefer interface to class, prefer interface/class higher in hierarchy than specific one. This is the basics of the OOD. 
Specifically here. `FireboltException` extends `SQLException` and therefore can be thrown from any  function that is declared as `throws SQLException`. The opposite is incorrect. If your function is declared as `throws FireboltException` it cannot throw `SQLException`. This is the reason that there many  places in code that look like the following:

**Review by ptiurin — APPROVED** [review]
LGTM, thanks for the change.

**Comment by alexradzin** [comment]
This historical PR has been replaced by https://github.com/firebolt-db/jdbc/pull/252

**Pull request closed without merging**
